src/ol_infrastructure/applications/xpro/__main__.py

```diff
@@ -0,0 +1,203 @@
+                        f"arn:aws:s3:::{xpro_storage_bucket_name}/*",
+                    ],
+                },
+            ],
+        },
+        stringify=True,
+        parliament_config={
+            "PERMISSIONS_MANAGEMENT_ACTIONS": {
+                "ignore_locations": [{"actions": ["s3:putobjectacl"]}]
+            },
+            "RESOURCE_EFFECTIVELY_STAR": {},
+        },
+    ),
+)
+
+xpro_vault_backend_role = vault.aws.SecretBackendRole(
+    "xpro-app",
+    name="xpro-app",
+    backend="aws-mitx",
+    credential_type="iam_user",
+    policy_arns=[xpro_iam_policy.arn],
+)
+
+# Create RDS instance
+xpro_db_security_group = ec2.SecurityGroup(
+    f"xpro-db-access-{stack_info.env_suffix}",
+    description=f"Access control for the xpro DB in {stack_info.name}",
+    ingress=[
+        ec2.SecurityGroupIngressArgs(
+            protocol="tcp",
+            from_port=DEFAULT_POSTGRES_PORT,
+            to_port=DEFAULT_POSTGRES_PORT,
+            cidr_blocks=["0.0.0.0/0"],
+            ipv6_cidr_blocks=["::/0"],
+            description="Allow access over the public internet from Heroku",
+        ),
+        ec2.SecurityGroupIngressArgs(
+            protocol="tcp",
+            from_port=DEFAULT_POSTGRES_PORT,
+            to_port=DEFAULT_POSTGRES_PORT,
+            security_groups=[data_vpc["security_groups"]["integrator"]],
+        ),
+    ],
+    tags=aws_config.merged_tags(
+        {"name": f"xpro-db-access-applications-{stack_info.env_suffix}"}
+    ),
+    vpc_id=apps_vpc["id"],
+)
+
+xpro_db_config = OLPostgresDBConfig(
+    instance_name=f"xpro-db-applications-{stack_info.env_suffix}",
+    password=xpro_config.require("db_password"),
+    subnet_group_name=apps_vpc["rds_subnet"],
+    security_groups=[xpro_db_security_group],
+    tags=aws_config.tags,
+    db_name="xpro",
+    public_access=True,
+    **defaults(stack_info)["rds"],
+)
+xpro_db = OLAmazonDB(xpro_db_config)
+
+xpro_vault_backend_config = OLVaultPostgresDatabaseConfig(
+    db_name=xpro_db_config.db_name,
+    mount_point=f"{xpro_db_config.engine}-xpro",
+    db_admin_username=xpro_db_config.username,
+    db_admin_password=xpro_db_config.password.get_secret_value(),
+    db_host=xpro_db.db_instance.address,
+    **defaults(stack_info)["rds"],
+)
+xpro_vault_backend = OLVaultDatabaseBackend(xpro_vault_backend_config)
+
+######################
+# Secrets Management #
+######################
+xpro_secrets = vault.Mount(
+    "xpro-vault-secrets-storage",
+    path="secret-xpro",
+    type="kv",
+    options={"version": 2},
+    description="Static secrets storage for the xpro application",
+    opts=ResourceOptions(delete_before_replace=True),
+)
+
+xpro_vault_secrets = read_yaml_secrets(
+    Path(f"xpro/secrets.{stack_info.env_suffix}.yaml")
+)
+
+for key, data in xpro_vault_secrets.items():
+    vault.kv.SecretV2(
+        f"xpro-vault-secrets-{key}",
+        name=key,
+        mount=xpro_secrets,
+        data_json=json.dumps(data),
+    )
```

Comment:
I like the one-key approach for storing stuff in vault. It is a little less thinking at the pulumi-stack stage.

This [file](https://github.com/mitodl/ol-infrastructure/blob/main/src/bridge/secrets/odl_video_service/data.qa.yaml) gets [loaded](https://github.com/mitodl/ol-infrastructure/blob/main/src/ol_infrastructure/applications/odl_video_service/__main__.py#L634-L638) into all under one key at this [location](https://vault-qa.odl.mit.edu/ui/vault/secrets/secret-odl-video-service/kv/ovs-secrets/details?version=14) so there is just one json object stored in vault containing everything about everything for OVS. Then all my calls to vault from my consul-template [look very similar](https://github.com/mitodl/ol-infrastructure/blob/main/src/bilder/images/odl_video_service/files/.env.tmpl#L19-L50).

But to each his own, I say. Multiple keys is probably better practice. 👍

Reply:
Thanks for commenting.

I don't have enough experience to form an opinion here.

Let me know if you think we should have a discussion around defining best practice, and we can put it in the patterns cookbook.

Reply:
I think that it's better to logically group secrets by key-path within the mount. That simplifies the work of permissions-management and versioning those logical elements separately.